File: ChatGPT3_5/State/round3/Code.py
from abc import ABC, abstractmethod

# State interface (abstract class)
class State(ABC):
    @abstractmethod
    def turn_on(self, light):
        """Handle turning the light on."""
        pass

    @abstractmethod
    def turn_off(self, light):
        """Handle turning the light off."""
        pass

# Concrete state: Light is ON
class OnState(State):
    def turn_on(self, light):
        """Handle case when light is already on."""
        print("The light is already on.")

    def turn_off(self, light):
        """Turn off the light and transition to OffState."""
        print("Turning the light off.")
        light.set_state(OffState())

# Concrete state: Light is OFF
class OffState(State):
    def turn_on(self, light):
        """Turn on the light and transition to OnState."""
        print("Turning the light on.")
        light.set_state(OnState())

    def turn_off(self, light):
        """Handle case when light is already off."""
        print("The light is already off.")

# Context class: Light
class Light:
    def __init__(self):
        # Initialize the light in the OffState
        self.state = OffState()

    def set_state(self, state):
        """Set the current state of the light."""
        self.state = state

    def turn_on(self):
        """Delegate to the state's turn_on method."""
        print("Attempting to turn on the light...")
        self.state.turn_on(self)

    def turn_off(self):
        """Delegate to the state's turn_off method."""
        print("Attempting to turn off the light...")
        self.state.turn_off(self)

File: Java_Gemini_Flash/State/round3/Code.java
package State.round3;

class Light {

    private State state;

    public Light() {
        state = new OffState();
    }

    public void turnOn() {
        state.turnOn(this);
    }

    public void turnOff() {
        state.turnOff(this);
    }

    public void setState(State state) {
        this.state = state;
    }

    public State getState() {
        return state;
    }

    public static void main(String[] args) {
        Light light = new Light();
        light.turnOn();
        light.turnOff();
    }
}

interface State {
    void turnOn(Light light);
    void turnOff(Light light);
}

class OnState implements State {

    @Override
    public void turnOn(Light light) {
        System.out.println("Light is already on!");
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Turning off the light...");
        light.setState(new OffState());
    }
}

class OffState implements State {

    @Override
    public void turnOn(Light light) {
        System.out.println("Turning on the light...");
        light.setState(new OnState());
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Light is already off!");
    }
}
